### PR TITLE
geoindex: begin a lib for planning index-accelerated geospatial queries

### PR DESCRIPTION
**Pull request by danhhz**
Release note: None

### CODE REVIEW

File: pkg/geo/geofn.go [hunk 1]
```diff
@@ -39,3 +41,75 @@ func STAreaGeography(g *Geography) (float64, error) {
...
+	}
+	switch f1 := sf1.(type) {
+	case s2.Point:
+		switch f2 := sf2.(type) {
+		case s2.Point:
+			return f1.ContainsPoint(f2), nil
+		case *s2.Polygon:
+			// TODO(dan): Pretty sure this is wrong if polygon is of size 0.
+			return false, nil
+		}
```
Comment (otan) [inline pkg/geo/geofn.go]: i'm guessing we're not worrying about LineString yet :) ?

File: pkg/geo/geofn.go [hunk 2]
```diff
@@ -39,3 +41,75 @@ func STAreaGeography(g *Geography) (float64, error) {
...
+	if err != nil {
+		return false, err
+	}
+	switch f1 := sf1.(type) {
+	case s2.Point:
+		switch f2 := sf2.(type) {
+		case s2.Point:
+			return f1.ContainsPoint(f2), nil
+		case *s2.Polygon:
+			// TODO(dan): Pretty sure this is wrong if polygon is of size 0.
```
Comment (otan) [inline pkg/geo/geofn.go]: you can't have a 0 point polygon i believe. polygons have to be closed. see 4.3.5: https://postgis.net/docs/using_postgis_dbmanagement.html

File: pkg/geo/geofn.go [hunk 3]
```diff
@@ -39,3 +41,75 @@ func STAreaGeography(g *Geography) (float64, error) {
...
+		case *s2.Polygon:
+			return f1.Contains(f2), nil
+		}
+	}
+	return false, errors.AssertionFailedf(
+		`unhandled feature types %T %T: %v %v`, g1.Figure, g2.Figure, g1, g2)
+}
+
+// STAsGeoJSONGeography exactly matches the semantics of ST_AsGeoJSON.
+func STAsGeoJSONGeography(g *Geography, properties map[string]interface{}) ([]byte, error) {
```
Comment (otan) [inline pkg/geo/geofn.go]: any reason not to have this use the `geom` library?

File: pkg/geo/geo.go [hunk 4]
```diff
@@ -41,6 +41,11 @@ type Geography struct {
 	SRID   geopb.SRID
 }
 
+// NewGeography constructs a new Geography with the given figure and SRID.
+func NewGeography(srid geopb.SRID, figure ...s2.Region) *Geography {
```
Comment (otan) [inline pkg/geo/geo.go]: wondering if we want to make this return a non-ptr version and call this `MakeGeography` instead....

**Review by otan — COMMENTED** [review]
going to look at the indexing later, some top level comments first!

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46051)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geo.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2FRVB3zv1pwxaAfHrm-r1-45:-M2ZpBaw8HZmG237NqJ_:b64iy7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5232f346bcebd5956b19c48f8b32ed2cd4aeddb/pkg/geo/geo.go#L45)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

wondering if we want to make this return a non-ptr version and call this `MakeGeography` instead....
</blockquote></details>

When something is usually used as a pointer (as Geography is here, all the methods are on `*Geography`) then we do New and return a pointer. The Make variant is only useful if it's going to let us save allocs
___
*[pkg/geo/geofn.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2FRVB4QLK2rhVYSTQZ-r1-61:-M2ZokmF0yfOhOVQ3bA4:b-m7zv52) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5232f346bcebd5956b19c48f8b32ed2cd4aeddb/pkg/geo/geofn.go#L61)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

you can't have a 0 point polygon i believe. polygons have to be closed. see 4.3.5: https://postgis.net/docs/using_postgis_dbmanagement.html
</blockquote></details>

Yup, you're right
___
*[pkg/geo/geofn.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2FRVB4QLK2rhVYSTQZ-r1-63:-M2Zomhg6zihmLB6IPgl:bufhn0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5232f346bcebd5956b19c48f8b32ed2cd4aeddb/pkg/geo/geofn.go#L63)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i'm guessing we're not worrying about LineString yet :) ?
</blockquote></details>

Yeah, lots to flesh out. Just filling in a skeleton for now. If anything changes, it'll be less mechanical work to refactor with fewer things.
___
*[pkg/geo/geofn.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2FRVB4QLK2rhVYSTQZ-r1-77:-M2ZovusBSjuJnEUpBec:b-ltj1qh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5232f346bcebd5956b19c48f8b32ed2cd4aeddb/pkg/geo/geofn.go#L77)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

any reason not to have this use the `geom` library?
</blockquote></details>

That only works for `Geometry`s right? We'd have to convert it first and ST_AsGeoJSON does have a Geography impl in PostGIS

**Review by otan — COMMENTED** [review]
how thoroughly should i be reviewing this prototype work? are we merging this straight onto master later?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-:-M2dVx28CEGSxB7lYV4s:b86lin1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @otan)

___
*[pkg/geo/geofn.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2FRVB4QLK2rhVYSTQZ-r1-77:-M2dTDqG3Z08VSo8UFls:b-j7icd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5232f346bcebd5956b19c48f8b32ed2cd4aeddb/pkg/geo/geofn.go#L77)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

That only works for `Geometry`s right? We'd have to convert it first and ST_AsGeoJSON does have a Geography impl in PostGIS
</blockquote></details>

hmm i guess so. 
___
*[pkg/geo/geofn.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2dTehe1xwqPs-pQq8b:-M2dTehe1xwqPs-pQq8c:b-7umc8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e5ca9dd827cff8e9c175fa0cc4119a9d268f971/pkg/geo/geofn.go#L108)):*
> ```Go
> 			}
> 		}
> 		feature.Geometry.Type = `Polygon`
> ```

so this could be multipolygon even if there's 1 loop. i *think* if we're using S2 encoding we have to add some metadata (srid, figuretype). please add a TODO
___
*[pkg/geo/geoindex/s2index.go, line 23 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2dV2J14SUGq01Qc6Iw:-M2dV2J14SUGq01Qc6Ix:b-f1fffg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e5ca9dd827cff8e9c175fa0cc4119a9d268f971/pkg/geo/geoindex/s2index.go#L23)):*
> ```Go
> // decomposition.
> //
> // TODO(dan): Explain how this works.
> ```

can you do this now? helpful for review purposes :D 
___
*[pkg/geo/geoindex/s2index.go, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2dVEI8E-dVc7ybqIrC:-M2dVEI8E-dVc7ybqIrD:bwzi0k7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e5ca9dd827cff8e9c175fa0cc4119a9d268f971/pkg/geo/geoindex/s2index.go#L52)):*
> ```Go
> 		querySpans[idx].End = Key(queryCellID.RangeMax())
> 	}
> 	regionArea, err := geo.STAreaGeography(g)
> ```

hmm, i wonder how badly this may get affected by spheroid mathematics...
___
*[pkg/geo/geoindex/s2index_test.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46051#-M2dVuxs05gImivbORjq:-M2dVuxs05gImivbORjr:b32vju) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e5ca9dd827cff8e9c175fa0cc4119a9d268f971/pkg/geo/geoindex/s2index_test.go#L31)):*
> ```Go
> func TestS2Index(t *testing.T) {
> 	rng, _ := randutil.NewPseudoRand()
> 	// rng.Seed(0)
> ```

:o

**Pull request closed without merging**